src/main/java/frc/trigon/robot/poseestimation/robotposesources/RobotPoseSource.java

```diff
@@ -76,11 +66,64 @@ public double getLastResultTimestamp() {
         return inputs.lastResultTimestamp;
     }
 
-    private Pose2d getUnCachedRobotPose() {
-        final Pose3d cameraPose = doubleArrayToPose3d(inputs.cameraPose);
+
+    /**
+     * Calculates the standard deviations of the estimated robot pose.
+     *
+     * @return the standard deviations for the pose estimation strategy used
+     */
+    public Matrix<N3, N1> calculateStdDevs() {
+        if (isWithinTagRange())
+            return calculateSolvePNPStdDevs();
+        return calculateAssumedHeadingStdDevs();
+    }
+
+    /**
+     * If the robot is close enough to the tag, it gets the estimated solve PNP pose.
+     * If it's too far, it assumes the robot's heading and calculates its position from there.
+     * Assuming the robot's heading is more robust, but using solve PNP resets the robot's heading.
+     *
+     * @param distanceFromBestTag the average of the distance from the best visible tag
+     * @return the robot's pose
+     */
+    private Pose2d calculateBestRobotPose(double distanceFromBestTag) {
+        if (isWithinTagRange())
+            return cameraPoseToRobotPose(inputs.solvePNPPose);
+        return calculateAssumedRobotHeadingPose().toPose2d();
+    }
+
+    /**
+     * Calculates the robot's pose by assuming its heading and the apriltag's position.
```

Comment:
This assumes another thing :)

Reply:
Well, I cant find it

Reply:
Look deeper 
Anything that isn't an input is assumed